chainer/links/connection/lstm.py

```diff
@@ -197,6 +211,11 @@ def __call__(self, x):
             ~chainer.Variable: Outputs of updated LSTM units.
 
         """
+        if self.upward.has_uninitialized_params:
+            in_size = x.size // len(x.data)
+            self.upward._initialize_params(in_size)
```

Comment:
In the original implementation of the LSTM link, the LSTM link must use its initializers after Linear links are initialized. Specifically, we first initialize weight matrices with temporal initializers to create ones and initialize them with genuine initializers afterwards. We follow this seemingly redundant procedure because passing `lateral_init` and `upward_init` directly to `Linear` links inside leads to wrong initialization (see the discussion [here](https://github.com/pfnet/chainer/pull/1086#r62287522)).

On the other hand, initializations should be done before concrete matrix multiplications, of course. So, the LSTM link must use its initializers between initialization by Linear links and concrete calculations. But as these are coupled in `Linear.__call__`, we cannot insert it between the two. So, we have to call `upward`'s initialization directly here.

I would like to know if there is more clever way of initialization.